src/planner/binder.cpp

```diff
@@ -516,11 +516,26 @@ void Binder::AddTableName(string table_name) {
 	root_binder.table_names.insert(std::move(table_name));
 }
 
+void Binder::AddReplacementScan(const string &table_name, unique_ptr<TableRef> replacement) {
+	auto &root_binder = GetRootBinder();
+	auto it = root_binder.replacement_scans.find(table_name);
+	if (it == root_binder.replacement_scans.end()) {
+		root_binder.replacement_scans[table_name] = std::move(replacement);
+	} else {
+		// A replacement scan by this name was previously registered, we can just use it
```

Comment:
A result of this is that aliases applied to the replacement scan are also cached, possibly causing some surprising behavior:
```sql
            SELECT Number from df t(Number)
            UNION ALL
            SELECT Number + (select Number from df offset 2 limit 1) + 1 as new
```
Because `df` is aliased as `t(Number)` the first time it's referenced, it now implicitly gains that alias in every future reference: `select Number from df`

Reply:
I wonder if I can just strip the aliases, it should be applied in the relevant locations anyways even if I strip it in the CTE definition we create

Reply:
Yea that worked, had to push the `column_name_alias` up to the TableRef out of the derived classes, which I think is only a plus

Reply:
I might need to also change the `alias` of the TableRef, testing now